File: apps/desktop/src/lib/index-overview.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/;
const COUNT_PATTERN = /^(?:0|[1-9][0-9]{0,19})$/;
const ZERO = BigInt(0);
const MAX_U64 = BigInt('18446744073709551615');
const MAX_PATH_DISPLAY_CHARS = 512;
const MAX_DIAGNOSTIC_FILES = 64;
const MAX_DIAGNOSTICS_PER_FILE = 8;
const MAX_DIAGNOSTIC_MESSAGE_BYTES = 1_024;
const MAX_U32 = 4_294_967_295;

export interface QueryIndexOverviewRequestV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type IndexLanguageV1 = 'generic' | 'rust' | 'typeScriptJavaScript' | 'python';
export type IndexDiagnosticCodeV1 =
  'syntaxError' | 'missingSyntax' | 'invalidEncoding' | 'unsupportedSyntax' | 'outputTruncated';
export type IndexDiagnosticSeverityV1 = 'error' | 'warning' | 'information';

export interface IndexDiagnosticV1 {
  code: IndexDiagnosticCodeV1;
  endByte: number;
  message: string;
  severity: IndexDiagnosticSeverityV1;
  startByte: number;
}

export interface IndexFileDiagnosticsV1 {
  coverageBasisPoints: number | null;
  diagnosticCount: string;
  diagnostics: IndexDiagnosticV1[];
  diagnosticsTruncated: boolean;
  language: IndexLanguageV1;
  pathDisplay: string;
  pathDisplayTruncated: boolean;
}

export interface IndexOverviewCountsV1 {
  diagnosticCount: string;
  diagnosticFileCount: string;
  fileCount: string;
  parsedFileCount: string;
  symbolCount: string;
}

export interface IndexOverviewV1 {
  counts: IndexOverviewCountsV1;
  coverageBasisPoints: number | null;
  diagnosticFiles: IndexFileDiagnosticsV1[];
  diagnosticFilesTruncated: boolean;
  snapshotId: string;
}

export type IndexOverviewResultV1 =
  | { status: 'noProject' }
  | { status: 'noPublishedIndex' }
  | { overview: IndexOverviewV1; status: 'published' };

export interface IndexOverviewResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: IndexOverviewResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function queryIndexOverview(
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<IndexOverviewResponseV1> {
  const request: QueryIndexOverviewRequestV1 = { protocolVersion: CURRENT_PROTOCOL_VERSION };
  const payload = await invokeCommand('query_index_overview', { request });
  return parseIndexOverviewResponseV1(payload);
}

export function parseIndexOverviewResponseV1(payload: unknown): IndexOverviewResponseV1 {
  if (
    !isRecord(payload) ||
    !hasExactKeys(payload, ['protocolVersion', 'result']) ||
    payload.protocolVersion !== CURRENT_PROTOCOL_VERSION
  ) {
    throw new Error('Index overview response does not match the V1 schema.');
  }
  return { protocolVersion: payload.protocolVersion, result: parseResult(payload.result) };
}

function parseResult(value: unknown): IndexOverviewResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Index overview response contains an invalid result.');
  }
  if (value.status === 'noProject' && hasExactKeys(value, ['status'])) {
    return { status: 'noProject' };
  }
  if (value.status === 'noPublishedIndex' && hasExactKeys(value, ['status'])) {
    return { status: 'noPublishedIndex' };
  }
  if (value.status === 'published' && hasExactKeys(value, ['overview', 'status'])) {
    return { overview: parseOverview(value.overview), status: 'published' };
  }
  throw new Error('Index overview response contains an invalid result.');
}

function parseOverview(value: unknown): IndexOverviewV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'counts',
      'coverageBasisPoints',
      'diagnosticFiles',
      'diagnosticFilesTruncated',
      'snapshotId',
    ]) ||
    !isStableId(value.snapshotId) ||
    !isCoverage(value.coverageBasisPoints) ||
    !Array.isArray(value.diagnosticFiles) ||
    value.diagnosticFiles.length > MAX_DIAGNOSTIC_FILES ||
    typeof value.diagnosticFilesTruncated !== 'boolean'
  ) {
    throw new Error('Index overview response contains an invalid publication.');
  }
  const counts = parseCounts(value.counts);
  const diagnosticFiles = value.diagnosticFiles.map(parseDiagnosticFile);
  const fileCount = BigInt(counts.fileCount);
  const parsedFileCount = BigInt(counts.parsedFileCount);
  const diagnosticFileCount = BigInt(counts.diagnosticFileCount);
  const diagnosticCount = BigInt(counts.diagnosticCount);
  const visibleDiagnosticCount = diagnosticFiles.reduce(
    (total, file) => total + BigInt(file.diagnosticCount),
    ZERO,
  );
  if (
    parsedFileCount > fileCount ||
    diagnosticFileCount > fileCount ||
    diagnosticFileCount < BigInt(diagnosticFiles.length) ||
    value.diagnosticFilesTruncated !== diagnosticFileCount > BigInt(diagnosticFiles.length) ||
    (parsedFileCount === ZERO) !== (value.coverageBasisPoints === null) ||
    diagnosticCount < visibleDiagnosticCount ||
    (!value.diagnosticFilesTruncated && diagnosticCount !== visibleDiagnosticCount)
  ) {
    throw new Error('Index overview response contains contradictory aggregate values.');
  }
  return {
    counts,
    coverageBasisPoints: value.coverageBasisPoints,
    diagnosticFiles,
    diagnosticFilesTruncated: value.diagnosticFilesTruncated,
    snapshotId: value.snapshotId,
  };
}

function parseCounts(value: unknown): IndexOverviewCountsV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'diagnosticCount',
      'diagnosticFileCount',
      'fileCount',
      'parsedFileCount',
      'symbolCount',
    ]) ||
    !isCount(value.diagnosticCount) ||
    !isCount(value.diagnosticFileCount) ||
    !isCount(value.fileCount) ||
    !isCount(value.parsedFileCount) ||
    !isCount(value.symbolCount)
  ) {
    throw new Error('Index overview response contains invalid counts.');
  }
  return {
    diagnosticCount: value.diagnosticCount,
    diagnosticFileCount: value.diagnosticFileCount,
    fileCount: value.fileCount,
    parsedFileCount: value.parsedFileCount,
    symbolCount: value.symbolCount,
  };
}

function parseDiagnosticFile(value: unknown): IndexFileDiagnosticsV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, [
      'coverageBasisPoints',
      'diagnosticCount',
      'diagnostics',
      'diagnosticsTruncated',
      'language',
      'pathDisplay',
      'pathDisplayTruncated',
    ]) ||
    !isCoverage(value.coverageBasisPoints) ||
    !isCount(value.diagnosticCount) ||
    BigInt(value.diagnosticCount) === ZERO ||
    !Array.isArray(value.diagnostics) ||
    value.diagnostics.length > MAX_DIAGNOSTICS_PER_FILE ||
    typeof value.diagnosticsTruncated !== 'boolean' ||
    !isLanguage(value.language) ||
    typeof value.pathDisplay !== 'string' ||
    value.pathDisplay.length === 0 ||
    Array.from(value.pathDisplay).length > MAX_PATH_DISPLAY_CHARS ||
    containsControl(value.pathDisplay) ||
    typeof value.pathDisplayTruncated !== 'boolean'
  ) {
    throw new Error('Index overview response contains invalid file diagnostics.');
  }
  const diagnostics = value.diagnostics.map(parseDiagnostic);
  const diagnosticCount = BigInt(value.diagnosticCount);
  if (
    diagnosticCount < BigInt(diagnostics.length) ||
    value.diagnosticsTruncated !== diagnosticCount > BigInt(diagnostics.length) ||
    (value.language === 'generic') !== (value.coverageBasisPoints === null)
  ) {
    throw new Error('Index overview response contains contradictory file diagnostics.');
  }
  return {
    coverageBasisPoints: value.coverageBasisPoints,
    diagnosticCount: value.diagnosticCount,
    diagnostics,
    diagnosticsTruncated: value.diagnosticsTruncated,
    language: value.language,
    pathDisplay: value.pathDisplay,
    pathDisplayTruncated: value.pathDisplayTruncated,
  };
}

function parseDiagnostic(value: unknown): IndexDiagnosticV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['code', 'endByte', 'message', 'severity', 'startByte']) ||
    !isDiagnosticCode(value.code) ||
    !isDiagnosticSeverity(value.severity) ||
    typeof value.message !== 'string' ||
    new TextEncoder().encode(value.message).length === 0 ||
    new TextEncoder().encode(value.message).length > MAX_DIAGNOSTIC_MESSAGE_BYTES ||
    containsControl(value.message) ||
    !isU32(value.startByte) ||
    !isU32(value.endByte) ||
    value.endByte < value.startByte
  ) {
    throw new Error('Index overview response contains an invalid diagnostic.');
  }
  return {
    code: value.code,
    endByte: value.endByte,
    message: value.message,
    severity: value.severity,
    startByte: value.startByte,
  };
}

function isCoverage(value: unknown): value is number | null {
  return (
    value === null ||
    (Number.isInteger(value) && typeof value === 'number' && value >= 0 && value <= 10_000)
  );
}

function isCount(value: unknown): value is string {
  return typeof value === 'string' && COUNT_PATTERN.test(value) && BigInt(value) <= MAX_U64;
}

function isStableId(value: unknown): value is string {
  return typeof value === 'string' && STABLE_ID_PATTERN.test(value);
}

function isU32(value: unknown): value is number {
  return Number.isInteger(value) && typeof value === 'number' && value >= 0 && value <= MAX_U32;
}

function isLanguage(value: unknown): value is IndexLanguageV1 {
  return (
    value === 'generic' ||
    value === 'rust' ||
    value === 'typeScriptJavaScript' ||
    value === 'python'
  );
}

function isDiagnosticCode(value: unknown): value is IndexDiagnosticCodeV1 {
  return (
    value === 'syntaxError' ||
    value === 'missingSyntax' ||
    value === 'invalidEncoding' ||
    value === 'unsupportedSyntax' ||
    value === 'outputTruncated'
  );
}

function isDiagnosticSeverity(value: unknown): value is IndexDiagnosticSeverityV1 {
  return value === 'error' || value === 'warning' || value === 'information';
}

function containsControl(value: string): boolean {
  return Array.from(value).some((character) => {
    const codePoint = character.codePointAt(0);
    return (
      codePoint !== undefined && (codePoint <= 0x1f || (codePoint >= 0x7f && codePoint <= 0x9f))
    );
  });
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/index_overview.rs
use crate::ProtocolVersion;
use serde::{Deserialize, Serialize};

/// Strict pathless input for the bounded V1 published-index overview.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct QueryIndexOverviewRequestV1 {
    protocol_version: ProtocolVersion,
}

impl QueryIndexOverviewRequestV1 {
    /// Creates a request for a specific protocol version.
    #[must_use]
    pub const fn new(protocol_version: ProtocolVersion) -> Self {
        Self { protocol_version }
    }

    /// Creates a request for the protocol version emitted by this build.
    #[must_use]
    pub const fn current() -> Self {
        Self::new(ProtocolVersion::CURRENT)
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(self) -> ProtocolVersion {
        self.protocol_version
    }
}

/// Versioned bounded result selected from the Core-owned active project.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexOverviewResponseV1 {
    protocol_version: ProtocolVersion,
    result: IndexOverviewResultV1,
}

impl IndexOverviewResponseV1 {
    /// Creates the response used before a project is active.
    #[must_use]
    pub const fn no_project() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: IndexOverviewResultV1::NoProject,
        }
    }

    /// Creates the response used before the first atomic publication.
    #[must_use]
    pub const fn no_published_index() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: IndexOverviewResultV1::NoPublishedIndex,
        }
    }

    /// Creates a response containing one bounded immutable publication projection.
    #[must_use]
    pub fn published(overview: IndexOverviewV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: IndexOverviewResultV1::Published {
                overview: Box::new(overview),
            },
        }
    }

    /// Returns the mutually exclusive active-project/publication result.
    #[must_use]
    pub const fn result(&self) -> &IndexOverviewResultV1 {
        &self.result
    }
}

/// Whether an active project and complete published index exist.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum IndexOverviewResultV1 {
    /// No project is active in this desktop process.
    NoProject,
    /// A project is active but no index has crossed the atomic publish boundary.
    NoPublishedIndex,
    /// The latest complete publication is available.
    Published {
        /// Bounded counts, coverage, and diagnostic-file summaries.
        overview: Box<IndexOverviewV1>,
    },
}

/// Bounded WebView-safe view of one published index snapshot.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexOverviewV1 {
    snapshot_id: String,
    counts: IndexOverviewCountsV1,
    coverage_basis_points: Option<u16>,
    diagnostic_files: Vec<IndexFileDiagnosticsV1>,
    diagnostic_files_truncated: bool,
}

impl IndexOverviewV1 {
    /// Creates a projection from application-validated bounded values.
    #[must_use]
    pub const fn new(
        snapshot_id: String,
        counts: IndexOverviewCountsV1,
        coverage_basis_points: Option<u16>,
        diagnostic_files: Vec<IndexFileDiagnosticsV1>,
        diagnostic_files_truncated: bool,
    ) -> Self {
        Self {
            snapshot_id,
            counts,
            coverage_basis_points,
            diagnostic_files,
            diagnostic_files_truncated,
        }
    }

    /// Returns the immutable publication snapshot digest.
    #[must_use]
    pub fn snapshot_id(&self) -> &str {
        &self.snapshot_id
    }

    /// Returns exact aggregate counts encoded as lossless decimal text.
    #[must_use]
    pub const fn counts(&self) -> &IndexOverviewCountsV1 {
        &self.counts
    }

    /// Returns byte-weighted structural coverage, if any file was structurally parsed.
    #[must_use]
    pub const fn coverage_basis_points(&self) -> Option<u16> {
        self.coverage_basis_points
    }

    /// Returns at most 64 canonical files that contain parser diagnostics.
    #[must_use]
    pub fn diagnostic_files(&self) -> &[IndexFileDiagnosticsV1] {
        &self.diagnostic_files
    }

    /// Returns whether further diagnostic files were omitted.
    #[must_use]
    pub const fn diagnostic_files_truncated(&self) -> bool {
        self.diagnostic_files_truncated
    }
}

/// Exact publication counters represented without JavaScript integer loss.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexOverviewCountsV1 {
    file_count: String,
    symbol_count: String,
    diagnostic_count: String,
    parsed_file_count: String,
    diagnostic_file_count: String,
}

impl IndexOverviewCountsV1 {
    /// Creates a group of canonical decimal counters.
    #[must_use]
    pub const fn new(
        file_count: String,
        symbol_count: String,
        diagnostic_count: String,
        parsed_file_count: String,
        diagnostic_file_count: String,
    ) -> Self {
        Self {
            file_count,
            symbol_count,
            diagnostic_count,
            parsed_file_count,
            diagnostic_file_count,
        }
    }

    /// Returns the complete published file count.
    #[must_use]
    pub fn file_count(&self) -> &str {
        &self.file_count
    }

    /// Returns the complete published symbol count.
    #[must_use]
    pub fn symbol_count(&self) -> &str {
        &self.symbol_count
    }

    /// Returns the complete published parser-diagnostic count.
    #[must_use]
    pub fn diagnostic_count(&self) -> &str {
        &self.diagnostic_count
    }

    /// Returns the number of structurally parsed files.
    #[must_use]
    pub fn parsed_file_count(&self) -> &str {
        &self.parsed_file_count
    }

    /// Returns the number of files containing at least one diagnostic.
    #[must_use]
    pub fn diagnostic_file_count(&self) -> &str {
        &self.diagnostic_file_count
    }
}

/// Safe file-local parser result surfaced without an authoritative path capability.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexFileDiagnosticsV1 {
    path_display: String,
    path_display_truncated: bool,
    language: IndexLanguageV1,
    coverage_basis_points: Option<u16>,
    diagnostic_count: String,
    diagnostics: Vec<IndexDiagnosticV1>,
    diagnostics_truncated: bool,
}

impl IndexFileDiagnosticsV1 {
    /// Creates one bounded file display and its safe diagnostic subset.
    #[must_use]
    pub const fn new(
        path_display: String,
        path_display_truncated: bool,
        language: IndexLanguageV1,
        coverage_basis_points: Option<u16>,
        diagnostic_count: String,
        diagnostics: Vec<IndexDiagnosticV1>,
        diagnostics_truncated: bool,
    ) -> Self {
        Self {
            path_display,
            path_display_truncated,
            language,
            coverage_basis_points,
            diagnostic_count,
            diagnostics,
            diagnostics_truncated,
        }
    }
}

/// Stable language names understood by the V1 index UI.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexLanguageV1 {
    /// No structural parser handled this file.
    Generic,
    /// Rust source parsed by the Rust adapter.
    Rust,
    /// TypeScript or JavaScript source parsed by the shared adapter.
    TypeScriptJavaScript,
    /// Python source parsed by the Python adapter.
    Python,
}

/// Safe parser diagnostic without source text.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct IndexDiagnosticV1 {
    code: IndexDiagnosticCodeV1,
    severity: IndexDiagnosticSeverityV1,
    message: String,
    start_byte: u32,
    end_byte: u32,
}

impl IndexDiagnosticV1 {
    /// Creates a diagnostic from already validated application evidence.
    #[must_use]
    pub const fn new(
        code: IndexDiagnosticCodeV1,
        severity: IndexDiagnosticSeverityV1,
        message: String,
        start_byte: u32,
        end_byte: u32,
    ) -> Self {
        Self {
            code,
            severity,
            message,
            start_byte,
            end_byte,
        }
    }
}

/// Stable parser-independent diagnostic code.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexDiagnosticCodeV1 {
    /// Parser emitted an explicit error node.
    SyntaxError,
    /// Parser inserted missing syntax.
    MissingSyntax,
    /// Required source text was not valid UTF-8.
    InvalidEncoding,
    /// Syntax exceeded the adapter's structural contract.
    UnsupportedSyntax,
    /// A bounded parser collection omitted additional artifacts.
    OutputTruncated,
}

/// User-visible parser-diagnostic severity.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum IndexDiagnosticSeverityV1 {
    /// Structural output is incomplete for this region.
    Error,
    /// Output remains useful with reduced precision.
    Warning,
    /// Informational parser observation.
    Information,
}

#[cfg(test)]
mod tests {
    use super::{
        IndexDiagnosticCodeV1, IndexDiagnosticSeverityV1, IndexDiagnosticV1,
        IndexFileDiagnosticsV1, IndexLanguageV1, IndexOverviewCountsV1, IndexOverviewResponseV1,
        IndexOverviewV1,
    };
    use serde_json::json;

    #[test]
    fn published_overview_has_a_strict_bounded_shape() -> Result<(), serde_json::Error> {
        let response = IndexOverviewResponseV1::published(IndexOverviewV1::new(
            "11".repeat(32),
            IndexOverviewCountsV1::new(
                "2".to_owned(),
                "3".to_owned(),
                "1".to_owned(),
                "1".to_owned(),
                "1".to_owned(),
            ),
            Some(8_000),
            vec![IndexFileDiagnosticsV1::new(
                "src/lib.rs".to_owned(),
                false,
                IndexLanguageV1::Rust,
                Some(8_000),
                "1".to_owned(),
                vec![IndexDiagnosticV1::new(
                    IndexDiagnosticCodeV1::SyntaxError,
                    IndexDiagnosticSeverityV1::Error,
                    "syntax error".to_owned(),
                    8,
                    10,
                )],
                false,
            )],
            false,
        ));

        assert_eq!(
            serde_json::to_value(response)?,
            json!({
                "protocolVersion": 1,
                "result": {
                    "status": "published",
                    "overview": {
                        "snapshotId": "11".repeat(32),
                        "counts": {
                            "fileCount": "2",
                            "symbolCount": "3",
                            "diagnosticCount": "1",
                            "parsedFileCount": "1",
                            "diagnosticFileCount": "1"
                        },
                        "coverageBasisPoints": 8000,
                        "diagnosticFiles": [{
                            "pathDisplay": "src/lib.rs",
                            "pathDisplayTruncated": false,
                            "language": "rust",
                            "coverageBasisPoints": 8000,
                            "diagnosticCount": "1",
                            "diagnostics": [{
                                "code": "syntaxError",
                                "severity": "error",
                                "message": "syntax error",
                                "startByte": 8,
                                "endByte": 10
                            }],
                            "diagnosticsTruncated": false
                        }],
                        "diagnosticFilesTruncated": false
                    }
                }
            })
        );
        Ok(())
    }
}
